README.md
# `openapi-interfaces`: Automatically generate `GET`, `POST`, `PUT` and JSON Merge Patch schemas for OpenAPI

**EXPERIMENTAL.** Details subject to change.

This tool extends [OpenAPI 3.1.0](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md) with the ability to generate related schemas automatically. Specifically, you can define a single interface type `Widget`, and automatically generate:

- `Widget` (for `GET` requests)
- `WidgetPost`
- `WidgetPut`
- `WidgetMergePatch` (for `PATCH` requests using [JSON Merge Patch](https://datatracker.ietf.org/doc/html/rfc7396) format, which is basically a nicely formalized "partial `PUT`")

## Installation

To install the latest version, first make sure you have a Rust toolchain. You
can [install one using these instructions](https://rustup.rs/). Then run:

```sh
cargo install -f openapi-interfaces
```

We will provide binaries at some point.

## Usage

```sh
openapi-interfaces --help
openapi-interfaces api_with_interfaces.yml -o api.yml
```

## OpenAPI extensions

This tool defines a new `components.interfaces` section, which allows specifying "interface" types. For example:

```yaml
components:
  # You can declare schemas normally if you wish.
  schemas: {}

  # But we also support interface definitions.
  interfaces:
    Resource:
      emit: false # Do not include this in generated output.
      members:
        id:
          required: true
          schema:
            # Normal OpenAPI / JSON Schema definitions.
            type: string
            format: uuid
    Widget:
      # Include all fields from `Resource`.
      $includes: "Resource"
      members:
        # We can override properties from `Resource` using JSON
        # Merge Patch syntax.
        id:
          schema:
            example: e35a3c8d-5486-49ec-9b23-6747afc19570
        name:
          required: true
          mutable: true
          schema:
            type: string
        comment:
          mutable: true
          schema:
            type: string
        readonly:
          required: true
          # This can't be updated once the object is created.
          mutable: false
          # But we do allow this to be set at creation time.
          # If omitted, `initializable` defaults to the value
          # of the `mutable` option.
          initializable: true
          schema:
            type: string
```

This will automatically generate four `Widget` types:

```yaml
components:
  schemas:
    Widget: ...
    WidgetPost: ...
    WidgetPut: ...
    WidgetMergePatch: ...
```

For the complete definitions, see [`example_output.yml`](./examples/example_output.yml).

### Referring to interfaces

We can then refer to interfaces using the new `$interface` key, with an appropriate variant:

```yaml
paths:
  /widgets:
    post:
      requestBody:
        required: true
        content:
          application/json:
              schema:
                # This becomes `$ref: "#/components/schemas/WidgetPost`.
                $interface: "Widget#Post"
      responses:
        201:
          content:
            application/json:
              schema:
                # This becomes `$ref: "#/components/schemas/Widget`.
                $interface: "Widget"
```

Possible options are:

- `Widget`: The type returned by `GET` and other methods that provide a complete resource from the server.
- `Widget#Post`: The type passed to a `POST` request.
- `Widget#Put`: The type passed to a `PUT` request. May also be used as the base type for applying `#MergePatch` values.
- `Widget#MergePatch`: A [JSON Merge Patch](https://datatracker.ietf.org/doc/html/rfc7396) schema that can be passed to `PATCH`.
- `Widget#SameAsInterface`: **May only be used inside `components.interfaces`.** This is a shortcut value that says "When generating a `#Get` interface, include `$ref: "components.schemas.Widget`. When generating a `#Post` interface, include `$ref: "components.schemas.WidgetPost`." In other words, use the same variant selector as the containing interface. Useful for when compound types are sent over the wire.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.4.0 - 2022-06-13

### Changed

- `--avoid-type-null` is now deprecated, and acts as an alias for `--use-nullable-for-merge-patch`. This flag now generates complete OpenAPI 3.0-compatible MergePatch types using a clever hack based on `nullable: true` and `oneOf:`. It also now sets the file version number to `openapi: 3.0.0` when used, because 3.1 does not support `nullable: true`.

## 0.3.0 - 2022-06-13

### Added

- Stable release. See [`CHANGELOG.md`](https://github.com/faradayio/openapi-interfaces/blob/main/CHANGELOG.md) for details of fixes and features since 0.2.1.

### Changed

- See the "Changed" section in v0.2.2-beta.4. This should only affect people using `discriminator` and `oneOf` interfaces.

## 0.3.0-rc.3 - 2022-06-10

### Fixed

- Include the discriminator in `Put` types, as originally planned (#31).

## 0.3.0-rc.2 - 2022-06-09

### Fixed

- Fix bug where we generated spurious `type: ~` entries for certain schema types using `oneOf` (#30).

## 0.3.0-rc.1 - 2022-06-08

### Added

- Binaries for Linux and Mac systems ([#27](https://github.com/faradayio/openapi-interfaces/issues/27)).
- Allow including definitions from external files using `$includeFiles` ([#24](https://github.com/faradayio/openapi-interfaces/issues/24)). See [`examples/include_files/`](https://github.com/faradayio/openapi-interfaces/tree/main/examples/include_files) for an example of how this works. Start from `base.yml`.

### Changed

- See the "Changed" section in v0.2.2-beta.4. This should only affect people using `discriminator` and `oneOf` interfaces.

### Fixed

- Improved error messages in many cases ([#28](https://github.com/faradayio/openapi-interfaces/issues/28)).

## 0.2.2-beta.5 - 2022-06-03

### Fixed

- Published release is made from the public `main` branch.

## 0.2.2-beta.4 - 2022-06-03 [YANKED]

### Fixed

- Copy the `description` of an interface to all the variants we generate ([#23](https://github.com/faradayio/openapi-interfaces/issues/23)). This will allow removing the `allOf` workaround that has been used with Readme.com.
- Emit `type: "object"` for interfaces using `oneOf` ([#13](https://github.com/faradayio/openapi-interfaces/issues/13)).

### Changed

- BREAKING: Completely overhaul how we handle `oneOf` interfaces and discriminators ([#18](https://github.com/faradayio/openapi-interfaces/issues/18)). This work is ongoing and not yet fully documented, but you can find a working example in [`examples/oneof_example.yml`](https://github.com/faradayio/openapi-interfaces/blob/main/examples/oneof_example.yml). This will require changing existing YAML files to use the new, TypeScript-like style. However, this feature is still **incomplete**. For ongoing discussion, see ([#26](https://github.com/faradayio/openapi-interfaces/issues/26)).

## 0.2.2-beta.3 - 2022-04-22

### Added

- Added support for `title` in interfaces.

### Fixed

- Pass through `description` in interfaces.
- Do not output `discriminator: null` if no discriminator is present.

### Fixed

- Pass through

## 0.2.2-beta.2 - 2022-04-16

### Added

- **Experimental:** Add `title` and `description` to the `oneOf` schemas introduced for `MergePatch` types. The goal here is to make auto-generated documentation prettier.

## 0.2.2-beta.1 - 2022-04-15

### Added

- **Experimental:** Add support for using `oneOf` in interfaces. See [`oneof_example.yml`](https://github.com/faradayio/openapi-interfaces/blob/main/examples/oneof_example.yml) and [`oneof_example_output.yml`](https://github.com/faradayio/openapi-interfaces/blob/main/examples/oneof_example_output.yml) for example input and output.

## 0.2.1 - 2022-04-12

### Added

- Allow a [`description` field](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#fixed-fields-19) as an optional sibling to`$interface` and `$ref`. This will be included in the generated `$ref` field.

## 0.2.0 - 2021-10-02

### Added

- If `--avoid-type-null` is passed, generate `MergePatch` types that with the schema `type: object`. This allows basic compatibility with OpenAPI 3.0, but it skips all merge patch validation.
- When run on OpenAPI 3.0 files, act as if `--avoid-type-null` were set.

### Changed

- BREAKING: Don't generate schemas that would only match the empty object. For example, if an object has no mutable fields, don't generate the `MergePatch` or `Put` variants.

## 0.1.3 - 2021-10-02

More fixes for 0.1.1. It looks like we need more regression tests for the corner cases.

### Fixed

- Fixed bug introduced in 0.1.1 that broke `$ref` and `$interface` in `MergePatch` types.

## 0.1.2 - 2021-10-02

### Fixed

- Fixed bug introduced in 0.1.1 that prevented using `$ref` and `$interface` in with array schemas.
- Cleaned up some error messages.

## 0.1.1 - 2021-10-01

### Added

- Added support for `components.responses`.
- Slightly improved error messages for interfaces and schemas.

## 0.1.0 - 2021-09-23

### Added

- Initial implementation of interface generation.


src/openapi/ref_or.rs
//! Support for values that might be replaced by `$ref` or `$interface`.

use std::{collections::BTreeMap, fmt::Debug};

use anyhow::format_err;
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use serde_json::Value;

use super::{
    interface::InterfaceVariant, serde_helpers::deserialize_enum_helper, Scope,
    Transpile,
};

/// Support for better error messages when parsing.
pub trait ExpectedWhenParsing {
    /// A string describing what type of value we expect when parsing a type
    /// that implements this interface.
    fn expected_when_parsing() -> &'static str;
}

/// Either a $ref, an $interface or a value of type T.
#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
#[serde(untagged)]
pub enum RefOr<T>
where
    T: Clone + Debug + Eq + PartialEq + Serialize,
{
    /// A reference to a component defined elsewhere. When working with OpenAPI,
    /// it is important _not_ to evaluate `$ref`, because OpenAPI code
    /// generators handle `$ref` specially.
    Ref(Ref),
    /// A new-style `$interface` reference. Analogous to `$ref`, but it works
    /// with interface types.
    InterfaceRef(InterfaceRef),
    /// A primitive value.
    Value(T),
}

impl<'de, T> Deserialize<'de> for RefOr<T>
where
    T: Clone
        + Debug
        + DeserializeOwned
        + Eq
        + ExpectedWhenParsing
        + PartialEq
        + Serialize,
{
    // Manually deserialize for slightly better error messages. See
    // https://github.com/faradayio/openapi-interfaces/issues/28 for the whole
    // horrifying story.
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        use serde_yaml::{Mapping, Value};

        // Parse it as raw YAML.
        let yaml = Mapping::deserialize(deserializer)?;

        // Helper to construct YAML hash keys.
        let yaml_str = |s| Value::String(String::from(s));

        // Look for `$includes` or `$ref`, otherwise fall back.
        if yaml.contains_key(&yaml_str("$ref")) {
            Ok(RefOr::Ref(deserialize_enum_helper::<D, _>(
                "$ref schema",
                Value::Mapping(yaml),
            )?))
        } else if yaml.contains_key(&yaml_str("$interface")) {
            Ok(RefOr::InterfaceRef(deserialize_enum_helper::<D, _>(
                "$interface schema",
                Value::Mapping(yaml),
            )?))
        } else {
            Ok(RefOr::Value(deserialize_enum_helper::<D, _>(
                &format!(
                    "$ref, $interface or {}",
                    <T as ExpectedWhenParsing>::expected_when_parsing(),
                ),
                Value::Mapping(yaml),
            )?))
        }
    }
}

impl<T> Transpile for RefOr<T>
where
    T: Clone + Debug + Eq + PartialEq + Serialize + Transpile<Output = T>,
{
    type Output = Self;

    fn transpile(&self, scope: &Scope) -> anyhow::Result<Self::Output> {
        match self {
            RefOr::InterfaceRef(r) => Ok(RefOr::Ref(r.transpile(scope)?)),
            RefOr::Ref(r) => Ok(RefOr::Ref(r.transpile(scope)?)),
            RefOr::Value(val) => Ok(RefOr::Value(val.transpile(scope)?)),
        }
    }
}

/// A `$ref` schema.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct Ref {
    /// Path to reference.
    #[serde(rename = "$ref")]
    target: String,

    /// An optional human-readable description which overrides whatever is in
    /// the referenced schema.
    ///
    /// See
    /// [the OpenAPI docs](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#fixed-fields-19).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    description: Option<String>,

    /// YAML fields we want to pass through blindly.
    #[serde(flatten)]
    unknown_fields: BTreeMap<String, Value>,
}

impl Ref {
    /// Construct a ref pointing to `target`.
    pub fn new<S: Into<String>>(target: S, description: Option<String>) -> Ref {
        Ref {
            target: target.into(),
            description,
            unknown_fields: Default::default(),
        }
    }
}

impl Transpile for Ref {
    type Output = Self;

    fn transpile(&self, _scope: &Scope) -> anyhow::Result<Self::Output> {
        if !self.unknown_fields.is_empty() {
            return Err(format_err!(
                "`$ref:` no unknown sibling values allowed in {:?}",
                self
            ));
        }

        Ok(self.clone())
    }
}

/// Our custom `$interface` schema. Analogous to `$ref`, but refers to a specifc
/// variant of an interface.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct InterfaceRef {
    /// Path to reference.
    #[serde(rename = "$interface")]
    pub target: String,

    /// An optional human-readable description which overrides whatever is in
    /// the referenced schema.
    ///
    /// See
    /// [the OpenAPI docs](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#fixed-fields-19).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    description: Option<String>,

    /// YAML fields we want to pass through blindly.
    #[serde(flatten)]
    unknown_fields: BTreeMap<String, Value>,
}

impl Transpile for InterfaceRef {
    /// `InterfaceRef` values transpile to regular `Ref` value.
    type Output = Ref;

    fn transpile(&self, scope: &Scope) -> anyhow::Result<Self::Output> {
        // This type is defined by us, so let's enforce this rule.
        if !self.unknown_fields.is_empty() {
            return Err(format_err!(
                "`$interface:` no unknown sibling values allowed in {:?}",
                self
            ));
        }

        // Figure out which interface variant to use.
        let target = transpile_interface_ref_to_ref(&self.target, scope)?;

        // Build our ref.
        Ok(Ref::new(target, self.description.clone()))
    }
}

/// Split the value of an `$interface` reference into a base name and a
/// fragment.
pub fn split_interface_ref(interface_ref: &str) -> (&str, Option<&str>) {
    let fragment_pos = interface_ref.find('#');
    if let Some(fragment_pos) = fragment_pos {
        (
            &interface_ref[..fragment_pos],
            Some(&interface_ref[fragment_pos..]),
        )
    } else {
        (interface_ref, None)
    }
}

/// Convert the value of an `$interface` to the body of a `$ref`.
pub fn transpile_interface_ref_to_ref(
    interface_ref: &str,
    scope: &Scope,
) -> Result<String, anyhow::Error> {
    let (base, fragment) = split_interface_ref(interface_ref);
    let fragment = fragment.unwrap_or("");
    let variety = if fragment == "#SameAsInterface" {
        // Get the interface variant from the surrounding scope.
        if let Some(variant) = scope.variant {
            variant
        } else {
            return Err(format_err!(
                "cannot use #SameAsInterface outside of a `components.interfaces` declaration"
            ));
        }
    } else {
        fragment.parse::<InterfaceVariant>()?
    };
    let target = format!(
        "#/components/schemas/{}{}",
        base,
        variety.to_schema_suffix_str()
    );
    Ok(target)
}

#[test]
fn transpile_interface_with_description() {
    let interface_ref = InterfaceRef {
        target: "Widget".to_string(),
        description: Some("hello".to_string()),
        unknown_fields: Default::default(),
    };

    let actual_ref = interface_ref.transpile(&Scope::default()).unwrap();

    let expected_ref = Ref {
        target: "#/components/schemas/Widget".to_string(),
        description: Some("hello".to_string()),
        unknown_fields: Default::default(),
    };

    assert_eq!(actual_ref, expected_ref);
}
